src/backend/allocator/dmabuf.rs

```diff
@@ -233,6 +253,23 @@ impl Dmabuf {
         WeakDmabuf(Arc::downgrade(&self.0))
     }
 
+    /// Presumably compatible device for buffer import
+    ///
+    /// This is inferred from client apis, as there is no kernel api or other guarantee this is correct,
+    /// so it should only be treated as a hint.
+    #[cfg(feature = "backend_drm")]
+    pub fn node(&self) -> Option<DrmNode> {
+        *self.0.node.lock().unwrap()
+    }
+
+    /// Sets or unsets any node set for this dmabuf (see [`Dmabuf::node`]).
+    ///
+    /// May alter behavior of other parts of smithay using this as a hint.
+    #[cfg(feature = "backend_drm")]
+    pub fn set_node(&self, node: impl Into<Option<DrmNode>>) {
+        *self.0.node.lock().unwrap() = node.into();
+    }
```

Comment:
Do we allow to change the node on `Dmabuf` to avoid more complex initialization in `create_dmabuf`/`dmabuf_imported` or because we have/expect a use-case for it?

Reply:
Multiple things:

Firstly as you pointed out, we can't really set this during initialization in the dmabuf protocol, because while we do know the device for v3 and lower (and potentially for v6 and higher), we don't really for v4/v5. We can communicate a different device via the target_device of a tranche but we don't know, what the client does. As such this would need more complex handling during `dmabuf_imported` to ask downstream for the node without providing a `Dmabuf`, given that we need the node to construct in the first place. Allowing downstream to just set this afterwards felt easier.


Secondly this is not really a property of the buffer (again no api to query this), but state added by smithay and potentially consumed by smithay (e.g. the MultiRenderer). Since it alters behaviour it felt appropriate to allow downstream to override this (maybe they know better in some cases?). Having a buffer stuck on `None` for other potential api, where we can't know (but downstream might) feels wrong.

I feel like I also should point out, that the MultiRenderer actively sets this, if it gets a buffer with a node set to `None` to avoid testing all nodes next time. So again, more state than a property.

Reply:
> I feel like I also should point out, that the MultiRenderer actively sets this, if it gets a buffer with a node set to None to avoid testing all nodes next time. So again, more state than a property.

Ah, missed that part. The rest would be solvable (in case there even is something to solve) by delaying the creation. But I agree that would greatly complicate the current API, so probably not worth it.

I am not totally happy with this, but I also don't see an easier way, so :+1: